src/SharpLearning.Backend.Cntk.Test/MnistCnnExampleTest.cs

```diff
@@ -92,12 +96,154 @@ public void MnistCnnTest()
                     }
 
                     var samplesSeenNextBatch = (i + 2) * minibatchSize;
-                    if (samplesSeenNextBatch >= epochSize)
+                    if (samplesSeenNextBatch >= numberOftrainingSamples)
                     {
                         readerTrain.Reset();
                     }
                 }
             }
+
+            // Test model.
+            var csharpError = MnistSimpleExampleTest.TestModelMnistLoader(z, device, mnist);
+            Trace.WriteLine($"Test Error: {csharpError}");
+
+            // Save model.
+            var modelPath = "cnn_mnist_csharp_loader.dnn";
+            z.Save(modelPath);
+
+            // Test loaded model.
+            var loadedModel = Function.Load(modelPath, device);
+            var loadedModelError = MnistSimpleExampleTest.TestModelMnistLoader(loadedModel, device, mnist);
+
+            Trace.WriteLine("Loaded Model Error: " + loadedModelError);
+            Assert.AreEqual(csharpError, loadedModelError, 0.00001);
+
+            // Test against python example.
+            var pythonError = 0.89;
+            Assert.AreEqual(pythonError, csharpError, 0.00001);
+        }
+
+        const string FeatureId = "features";
+        const string LabelsId = "labels";
+
+        [TestMethod]
+        public void MnistCnnTest_MinibatchSource()
+        {
+            // Set global data and device type. 
+            var dataType = CntkDataType.Float;
+            DeviceDescriptor device = DeviceDescriptor.CPUDevice;
```

Comment:
@nietras It was necesarry to use the CPU devide inorder to get derterministic behavior. I am not sure, but it seems that the nvidia implementation is not always deterministic. There is some information in this issue, but it seems to be regarding the MaxPooling layer, even though one of the CNTK devs mentions the "Nvidia convolution implementation":
[Non-deterministic MaxPooling layer](https://github.com/Microsoft/CNTK/issues/1690). This also discusses a "Deterministic flag", so maybe it is possible to setup the GPU to run in a deterministic mode. This would be good for testing.

Reply:
@mdabros wow that is really weird. How can max pooling be non-deterministic? Or even convolve? They mention it was fixed in 2.2. Anyway, I am fine with sticking to CPU for now.